server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -1042,15 +1057,16 @@ void runDelete(SnapshotDeleteListener listener) {
                 }, listener::onFailure));
             } else {
                 // Write the new repository data first (with the removed snapshot), using no shard generations
+                final var saferListener = wrapWithWeakConsistencyProtection(listener);
```

Comment:
I wonder whether we should have the `wrapWithWeakConsistencyProtection` method takes an `IndexVersion` parameter so that the implementation can decide whether to wrap it accordingly. This would be closer to what `S3Repository` is doing right now. Since the wrapping logic is from the subclasses, I feel it is better to have them decide whether to wrap as well.

Assuming we make the above change, I would not mind sending the wrapped listener to a separate method to avoid having both listener and wrapper listener in a single method, e.g. something like
```java
void runDelete(listener) {
    doRunDelete(maybeWrapWithWeakConsistencyProtection(repositoryFormatIndexVersion, listener));
}
```

Reply:
Yes extracting methods for the two branches is a reasonable idea, done in 984267c3c0ae694c8d7961d34ea9a79072b9b25e.

However I think this is the better place to make the decision about this protection. `BlobStoreRepository` holds all the logic about the layout of blobs in the repository and therefore knows whether it's doing something that needs extra protection against weakly-consistent blob stores or not. If we were to add support for another weakly-consistent blob store, we would have the same condition in there.